### DISCUSSION TITLE
Could we use our own signed certificate access proper Kafka resources

### DISCUSSION THREAD

**Discussion by barryzhounb** [strimzi/strimzi-kafka-operator#5581] [General]
Now I have replaced clients-ca-certs and client-ca with our own certificates, that means we use our own certificates as root CA for clients-ca. 

We have an app that wants to connect Kafka with SSL mode. We sign app's CSR with our own certificates, then we obtain app's certificate and key that have been signed by our own certificates.

Under SSL mode, usually it needs to create a user and define ACL permissions, then this user can access connect Kafka via SSL and access proper resources. Now in our case, our app use its own signed certificates, doesn't use certificates that generated by Kafka user. My question is, could our app access proper Kafka resources?

**Comment by scholzj** [comment]
If you use the latest Strimzi version, you can use the `type: tls-external` authentication in the `KafkaUser` resource (assuming your certificates have the subject with `CN` only). You can also disable the user operator and just manage the ACLs or Quotas using the Kafka APIs of course.

**Reply by barryzhounb** [reply]
Hi @scholzj Thanks your reply.

Under TLS mode, when creating a new user, we can get user's certificates and key, then Kafka can authenticate against them. I am confused how Kafka knows this user and authorize this user to access specific topics (suppose we define user's ACL permission)? Does user's certificates/key contain some information related to user? Could you elaborate them? Thanks

**Reply by scholzj** [reply]
The certificates have something called subject. And the subject is used to build a username. So there is no need to link the individual certificates to user account (there is actually no way how to do it). You need to generate your CSR and sign it with the right subject.

The subject has a different fields which will combine into the user. CN (Common Name), O (organization), OU (Organization Unit)m country and more. When you use all of these, when you connect, the username will be something like this `CN=my-user,O=my-company,OU=my-deparment,C=CZ`.

You can in general use all these fields. But the user operator works only with the `CN` So if you want to use the user operator to manage the ACLs and Quotas with the `type: tls-external` authentication, you need to have only the `CN=` in your certificates subject.

If you want to use the other fields as well, you will need to manage things such as ACLs using the Kafka APIs directly yourself.

**Reply by barryzhounb** [reply]
Very clear, I will try it. Thanks expert @scholzj

**Reply by barryzhounb** [reply]
Hello @scholzj I have done verification for what you mentioned, it works well. Big thanks to you.

One more question to ask you. user's certificate is signed by clients-ca-cert, but in fact when user connects to Kafka broker, it is "Kafka broker" that perform authentication, kafka broker's root CA is cluster-ca-cert, I am confused for that? How Kafka broker authenticate against user who carries certificates that is signed by clients-ca-cert? Could you help elaborate it? Thanks.

**Reply by scholzj** [reply]
The Kafka brokers have the Clients CA certificate as well.

**Reply by barryzhounb** [reply]
You mean Kafka broker has add client-ca to truststore, correct?

**Reply by scholzj** [reply]
Yes, correct.

**Reply by barryzhounb** [reply]
make sense. thanks
